apps/plea/admin.py

```diff
@@ -193,9 +249,168 @@ def statistics_view(self, request):
+
+        initiation_types_by_case = {
+            a.case.initiation_type
+            for a in AuditEvent.objects.all()
+            if hasattr(a, "case") and hasattr(a.case, "initiation_type")
+        }
+        initiation_types_by_event_data = {
+            a.event_data["initiation_type"]
+            for a in AuditEvent.objects.all()
+            if hasattr(a, "event_data") and "initiation_type" in a.event_data
+        }
+
+        # Unique initiation_types for auditevents, with case initiation_types being preferred
+        initiation_types = initiation_types_by_event_data
+        initiation_types.update(initiation_types_by_case)
+
+        # Return the initiation_type to use as a filter parameter and the name
+        return list(set([
+            (
+                initiation_type,
+                [
+                    i[1]
+                    for i in INITIATION_TYPE_CHOICES
+                    if i[0] == initiation_type
+                ][0]
+            )
+            for initiation_type in initiation_types
+        ])) + [  # Special case; compound filter
+            (
+                "J|Q|S",
+                "SJPs, requisitions and summons",
+            )
+        ]
+
+    def queryset(self, request, queryset):
+        if self.value():
+            # Compound queries (could be done programatically if needed)
+            if self.value() == "J|Q|S":
+                return queryset.filter(case__initiation_type="J") | \
+                    queryset.filter(case__initiation_type="Q") | \
+                    queryset.filter(case__initiation_type="S") | \
+                    queryset.filter(event_data__contains={'initiation_type': "J"}) | \
+                    queryset.filter(event_data__contains={'initiation_type': "Q"}) | \
+                    queryset.filter(event_data__contains={'initiation_type': "S"})
+
+            # Filter on  particular initiation_type
+            else:
+                return queryset.filter(
+                    case__initiation_type=self.value()
+                ) | queryset.filter(
+                    event_data__contains={
+                        'initiation_type': self.value()
+                    }
+                )
+        else:
+            return queryset
+
+
+class AuditEventAdmin(admin.ModelAdmin):
+    """Audit events in the admin page"""
+
+    list_display = (
+        'id',
+        'event_datetime',
+        'event_subtype',
+        'initiation_type',
+        'case_link',
+    )
+    list_filter = (
+        "event_type",
+        "event_subtype",
+        AuditEventInitiationTypeFilter,
+        UrnFilter,
+    )
+    search_fields = ("case__urn", "event_data__urn")
+
+    def initiation_type(self, obj):
+        """Required for the admin_order_field"""
+        result = [
+            i[1]
+            for i in INITIATION_TYPE_CHOICES
+            if obj.case and obj.case.initiation_type == i[0]
+        ]
```

Comment:
This seems very similar to code in `CaseInitiationTypeFilter` and `AuditEventInitiationTypeFilter`. I wonder if it could be factored out and shared.